crates/kumod/src/smtp_server.rs

```diff
@@ -2264,6 +2280,177 @@ impl SmtpServerSession {
+        if self.socket.is_none() {
+            return Err(anyhow!("socket is none"));
+        }
+        tracing::trace!("reading proxy protocol header");
+        let header = loop {
+            let mut data = [0u8; 108];
+            tokio::select! {
+                _ = tokio::time::sleep(self.params.client_timeout) => {
+                    return Err(anyhow!("timeout reading proxy protocol header"));
+                }
+                size = self.socket.as_mut().unwrap().read(&mut data) => {
+                    match size {
+                        Err(err) => {
+                            tracing::trace!("error reading: {err:#}");
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Diagnostic {
+                                    level: Level::ERROR,
+                                    message: format!("error reading: {err:#}"),
+                                },
+                                when: Utc::now(),
+                            });
+                            return Err(anyhow!("error reading proxy protocol header: {err:#}"));
+                        }
+                        Ok(size) if size == 0 => {
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Diagnostic {
+                                    level: Level::ERROR,
+                                    message: "Peer Disconnected".to_string(),
+                                },
+                                when: Utc::now(),
+                            });
+                            return Err(anyhow!("peer disconnected while reading proxy protocol header"));
+                        }
+                        Ok(size) => {
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Read(data[0..size].to_vec()),
+                                when: Utc::now(),
+                            });
+                            self.read_buffer.extend_from_slice(&data[0..size]);
+                        }
+                    }
+                }
+                _ = self.shutdown.shutting_down() => {
+                    return Err(anyhow!("shutting down"));
+                }
+            };
+            let header = HeaderResult::parse(&self.read_buffer);
+            if header.is_complete() {
+                break header;
+            }
+        };
+
+        let mut addr: Option<IpAddr> = None;
+        let mut port: Option<u16> = None;
+        let mut dest_addr: Option<IpAddr> = None;
+        let mut dest_port: Option<u16> = None;
```

Comment:
See comment about clearing the buffer below

```suggestion
        let mut dest_port: Option<u16> = None;
        let mut consumed_bytes = 0;
```